src/middlewared/middlewared/plugins/nfs.py

```diff
@@ -403,19 +414,35 @@ async def validate(self, data, schema_name, verrors, old=None):
 
     @private
     def validate_paths(self, data, schema_name, verrors):
-        if osc.IS_LINUX:
-            # Ganesha does not have such a restriction, each path is a different share
-            return
-
+        locked_datasets = self.middleware.call_sync('zfs.dataset.locked_datasets')
```

Comment:
Why does it matter on share creation?

Reply:
It matters because we should not allow users to use a path which is a path of a locked dataset. i.e pool/test is locked, we will still have the path `pool/test` as the dataset was first umounted and we should not allow that path to be used because it is basically the path of a locked dataset.

Reply:
Well, we need this in FreeBSD to perform the check below, but why don't we allow to create such a share on SCALE?

Reply:
Why should we ? If the path being used is of a locked dataset, we should not be using it, at least not when it's enabled ?

Reply:
I don't see a problem in creating a share for locked dataset, it'll get locked immediately, and I'll unlock the dataset when it'll be convenient for me.

Reply:
We only do not allow it when the share is enabled. If the share is enabled and the dataset locked, we will not allow the share to be created

Reply:
They can still create it if they disable the share and use locked dataset's paths

Reply:
Well, I want to create it enabled so it starts working automatically when I unlock the dataset, why can't I do it?

Reply:
It will start working automatically as the dataset is unlocked even if it's disabled on share creation ? I am not sure I follow on why should we allow it to be created with it enabled and then issuing an alert that please disable the share ?

Reply:
I don't agree with `Please disable the share` either, I think it's better to say `the share won't work until you unlock the dataset`

Reply:
We can certainly change that. However please do let me know about this specific case and I'll make required changes then :)

Reply:
I still think we should allow creating a share that will get locked after creation, not only in NFS, everywhere. In NFS+FreeBSD we'll still need your check in order to be sure that we don't create share that uses two or more datasets simultaneously.

Reply:
We do allow creating a share that is using a locked dataset. We just don't allow it to be enabled. Can you please confirm this ? I just want to be sure we are on the same page

Reply:
The flow is when user is using a locked dataset:
1) If share is enabled we error out
2) If share is disabled, we allow creation to proceed

If we allow creation to complete in (1), i think it's not nice to start telling them that they should disable it as we won't work on it. Why did we allow them to create it while being enabled in the first place ?

As for unlock, if it's enabled/disabled, it will be enabled on dataset unlock

Reply:
I dont see the need of overcomplicating any of this.

If the dataset is locked simply dont allow the share to be created.

Reply:
@william-gr that's why I wanted the NEP, there are arguable things here... Why do you want to prohibit creation of share for a locked dataset?

Reply:
I dont see anything arguable to be honest...

Why do you want to allow share creation of a locked dataset? whats the use case?

People create shares to be able to share things. Why complicate ourselves a great matter for little to no benefit? We have to be practical and simplify as much as possible.

This seems like a super edge/corner case.

Reply:
I want to be as simple as possible too. Forbidding to create share on a locked dataset is less simple because we need to validate this. Allowing, on the other hand, requires nothing from us: share will be created as usually and then locked immediately by etc generator script issuing an alert.

Use case to avoid is:
1) Boot NAS
2) Try to create share
3) Receive an error
4) *damn, I need to go to another UI section and unlock the dataset*
5) Re-save the same form
Another use case to avoid:
1) Edit share that uses locked dataset to change an unrelated field
2) Receive an error
3) *damn, I only wanted to change unrelated field, why do I have to disable this share now and then remember to enable it again when I unlock the dataset*?

Reply:
Just for clarity, I’ll again add this link https://github.com/freenas/freenas/pull/4967#discussion_r443727930. 

We do allow edits and creation of locked shares as long as the share is marked as disabled. Let’s please decide keeping this in mind :)

Reply:
Well, in that case, if its less code and less complicated to simply allow creation with locked datasets and everything will magically work out of the box, why not?

Or is there more to it?

Reply:
To my understanding there is not. User is able to create/edit locked shares ( as long as they are disabled ) and they don't have to enable it again as they unlock the dataset, it will be enabled for them automatically. If there is something else we should cater to, please let me know and I'll make sure we address those concerns.

Reply:
> and they don't have to enable it again as they unlock the dataset, it will be enabled for them automatically

Wait, so if I had the share that uses locked dataset disabled for some unrelated reason, it will be automatically enabled when I unlock the dataset?

Reply:
Yes, if they want to restart attached attachments, then yes, we toggle them for the user - the option is exposed in the UI

Reply:
Well, that's even worse.

To sum this up, IMO:
- "Enabled" checkbox should only serve for explicitly enabling/disabling shares by user
- User can create enabled share with locked datasets, it will have "locked" icon and won't be generated into config files
- Unlocking the dataset simply reloads services which had shares that were exporting that dataset

Reply:
To sum it up, you are saying we should not follow the behaviour of  `export` where we toggle attachments, but just restart related services on lock/unlock and make sure that the shares/task don't use locked resources ?

Reply:
Also I think that might imply to `export` as well, partially ( the motivation of not toggling )

Reply:
Yes. `export`'s implementation is not perfect (though it keeps track of which shares it has disabled and `import` only re-enables them), if I were to work on it now, I would so something like you're doing now.

Reply:
Agreed with @themylogin

Reply:
Awesome, I'll introduce required changes. Thank you @themylogin @william-gr